storage/objects: TestV4SignedURL failed

This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 942ee06cb82bce4290452c0c4d57d4328630482d
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/3a390478-bd9c-4c5b-a66f-a4d76e79ad46), [Sponge](http://sponge2/3a390478-bd9c-4c5b-a66f-a4d76e79ad46)
status: failed

Error since it didn't come up here:
```
=== RUN   TestV4SignedURL
    objects_test.go:447: object content = "<?xml version='1.0' encoding='UTF-8'?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Details>No such object: golang-samples-tests-4-signed-url-bucket-name/foo.txt</Details></Error>"; want "hello world"
--- FAIL: TestV4SignedURL (1.89s)
```

@BrennaEpp can you take this one?

--- a/storage/objects/objects_test.go
+++ b/storage/objects/objects_test.go
@@ -391,16 +391,13 @@ func TestKMSObjects(t *testing.T) {
 func TestV4SignedURL(t *testing.T) {
 	tc := testutil.SystemTest(t)
 	ctx := context.Background()
-	client, err := storage.NewClient(ctx)
-	if err != nil {
-		t.Fatalf("storage.NewClient: %v", err)
-	}
-	defer client.Close()
 
 	bucketName := tc.ProjectID + "-signed-url-bucket-name"
 	objectName := "foo.txt"
 
 	testutil.CleanBucket(ctx, t, tc.ProjectID, bucketName)
+
+	// Generate PUT URL.
 	putBuf := new(bytes.Buffer)
 	putURL, err := generateV4PutObjectSignedURL(putBuf, bucketName, objectName)
 	if err != nil {
@@ -411,17 +408,7 @@ func TestV4SignedURL(t *testing.T) {
 		t.Errorf("got %q, want %q", got, want)
 	}
 
-	httpClient := &http.Client{}
-	request, err := http.NewRequest("PUT", putURL, strings.NewReader("hello world"))
-	if err != nil {
-		t.Fatalf("failed to compose HTTP request: %v", err)
-	}
-	request.ContentLength = 11
-	request.Header.Set("Content-Type", "application/octet-stream")
-	_, err = httpClient.Do(request)
-	if err != nil {
-		t.Errorf("httpClient.Do: %v", err)
-	}
+	// Generate GET URL.
 	getBuf := new(bytes.Buffer)
 	getURL, err := generateV4GetObjectSignedURL(getBuf, bucketName, objectName)
 	if err != nil {
@@ -432,20 +419,37 @@ func TestV4SignedURL(t *testing.T) {
 		t.Errorf("got %q, want %q", got, want)
 	}
 
-	response, err := http.Get(getURL)
+	// Create PUT request.
+	httpClient := &http.Client{}
+	request, err := http.NewRequest("PUT", putURL, strings.NewReader("hello world"))
 	if err != nil {
-		t.Errorf("http.Get: %v", err)
+		t.Fatalf("failed to compose HTTP request: %v", err)
 	}
-	defer response.Body.Close()
+	request.ContentLength = 11
+	request.Header.Set("Content-Type", "application/octet-stream")
 
-	body, err := ioutil.ReadAll(response.Body)
-	if err != nil {
-		t.Errorf("ioutil.ReadAll: %v", err)
-	}
+	// Test PUT and GET requests.
+	testutil.Retry(t, 10, time.Second, func(r *testutil.R) {
+		_, err = httpClient.Do(request)
+		if err != nil {
+			t.Errorf("httpClient.Do: %v", err)
+		}
 
-	if got, want := string(body), "hello world"; got != want {
-		t.Errorf("object content = %q; want %q", got, want)
-	}
+		response, err := http.Get(getURL)
+		if err != nil {
+			t.Errorf("http.Get: %v", err)
+		}
+		defer response.Body.Close()
+
+		body, err := io.ReadAll(response.Body)
+		if err != nil {
+			t.Errorf("io.ReadAll: %v", err)
+		}
+
+		if got, want := string(body), "hello world"; got != want {
+			t.Errorf("object content = %q; want %q", got, want)
+		}
+	})
 }
 
 func TestPostPolicyV4(t *testing.T) {

--- a/storage/objects/objects_test.go
+++ b/storage/objects/objects_test.go
@@ -432,22 +432,22 @@ func TestV4SignedURL(t *testing.T) {
 	testutil.Retry(t, 10, time.Second, func(r *testutil.R) {
 		_, err = httpClient.Do(request)
 		if err != nil {
-			t.Errorf("httpClient.Do: %v", err)
+			r.Errorf("httpClient.Do: %v", err)
 		}
 
 		response, err := http.Get(getURL)
 		if err != nil {
-			t.Errorf("http.Get: %v", err)
+			r.Errorf("http.Get: %v", err)
 		}
 		defer response.Body.Close()
 
 		body, err := io.ReadAll(response.Body)
 		if err != nil {
-			t.Errorf("io.ReadAll: %v", err)
+			r.Errorf("io.ReadAll: %v", err)
 		}
 
 		if got, want := string(body), "hello world"; got != want {
-			t.Errorf("object content = %q; want %q", got, want)
+			r.Errorf("object content = %q; want %q", got, want)
 		}
 	})
 }

